pkg/koordlet/statesinformer/states_device.go

```diff
@@ -0,0 +1,212 @@
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/klog/v2"
+
+	"github.com/koordinator-sh/koordinator/apis/extension"
+	schedulingv1alpha1 "github.com/koordinator-sh/koordinator/apis/scheduling/v1alpha1"
+	"github.com/koordinator-sh/koordinator/pkg/koordlet/metriccache"
+)
+
+func generateQueryParam() *metriccache.QueryParam {
+	end := time.Now()
+	start := end.Add(-time.Duration(60) * time.Second)
+	return &metriccache.QueryParam{
+		Aggregate: metriccache.AggregationTypeLast,
+		Start:     &start,
+		End:       &end,
+	}
+}
+
+func (s *statesInformer) reportDevice() {
+	s.nodeRWMutex.RLock()
+	copyNode := s.node.DeepCopy()
+	s.nodeRWMutex.RUnlock()
+	gpuDevices := s.buildGPUDevice()
+	blocker := true
+	device := &schedulingv1alpha1.Device{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: copyNode.Name,
+			OwnerReferences: []metav1.OwnerReference{
+				{
+					APIVersion:         "v1",
+					Kind:               "Node",
+					Name:               copyNode.Name,
+					UID:                copyNode.UID,
+					Controller:         &blocker,
+					BlockOwnerDeletion: &blocker,
+				},
+			},
+		},
+		Spec: schedulingv1alpha1.DeviceSpec{
+			Devices: gpuDevices,
+		},
+	}
+	// TODO: compare diff before update
+	_, err := s.deviceClient.Update(context.TODO(), device, metav1.UpdateOptions{})
+	if err != nil {
+		if !errors.IsNotFound(err) {
+			klog.Errorf("failed to update device %s, err: %v", copyNode.Name, err)
+			return
+		}
+		_, err = s.deviceClient.Create(context.TODO(), device, metav1.CreateOptions{})
+		if err != nil {
+			klog.Errorf("failed to create device %s, err: %v", copyNode.Name, err)
+			return
+		}
+	}
+}
+
+func (s *statesInformer) buildGPUDevice() []schedulingv1alpha1.DeviceInfo {
+	queryParam := generateQueryParam()
+	nodeResource := s.metricsCache.GetNodeResourceMetric(queryParam)
+	if nodeResource.Error != nil {
+		klog.Errorf("failed to get node resource metric, err: %v", nodeResource.Error)
+		return nil
+	}
+	if len(nodeResource.Metric.GPUs) == 0 {
+		klog.V(5).Info("no gpu device found")
+		return nil
+	}
+	var deviceInfos []schedulingv1alpha1.DeviceInfo
+	for _, gpu := range nodeResource.Metric.GPUs {
+		health := true
+		s.gpuMutex.RLock()
+		if _, ok := s.unhealthyGPU[gpu.DeviceUUID]; ok {
+			health = false
+		}
+		s.gpuMutex.RUnlock()
+		deviceInfos = append(deviceInfos, schedulingv1alpha1.DeviceInfo{
+			UUID:   gpu.DeviceUUID,
+			Minor:  gpu.Minor,
+			Type:   schedulingv1alpha1.GPU,
+			Health: health,
+			Resources: map[corev1.ResourceName]resource.Quantity{
+				extension.GPUCore:        *resource.NewQuantity(100, resource.DecimalSI),
+				extension.GPUMemory:      gpu.MemoryTotal,
+				extension.GPUMemoryRatio: *resource.NewQuantity(100, resource.DecimalSI),
+			},
+		})
+	}
+	return deviceInfos
+}
+
+func (s *statesInformer) gpuHealCheck(stopCh <-chan struct{}) {
+	count, ret := nvml.DeviceGetCount()
```

Comment:
Call nvml.Init first. It will return an error if driver not found.

Reply:
https://github.com/koordinator-sh/koordinator/blob/main/pkg/koordlet/metricsadvisor/collector_gpu_linux.go#L57